### ISSUE TITLE
Regression on trunk compiling QCheck with PR 12389

### ISSUE THREAD

**Issue by jmid** [ocaml/ocaml#12543]
Since the merge of #12389 we're observing a regression which means [QCheck](https://github.com/c-cube/qcheck) fails to compile with
```
# Error: Could not find the .cmi file for interface src/core/QCheck2.ml.
```

The following works - with 44c9baa3ef representing the point just before the 12389 merge:
```
git checkout 44c9baa3ef
opam switch create . --empty
opam pin add -k path --inplace-build ocaml-variants.5.2.0+pre-12389 .
opam install qcheck-core
```

However the following now fails - with c2b87d8fad being the head of `trunk` since Friday:
```
git checkout c2b87d8fad
opam switch create . --empty
opam pin add -k path --inplace-build ocaml-variants.5.2.0+post-12389 .
opam install qcheck-core
```

The full error message is (with apologies for the noise from the unused field):
```
(cd _build/default && /home/jmi/software/ocaml-11-09-2023/_opam/bin/ocamlopt.opt -w -40 -w +a-4-42-44-48-50-58-32-60@8 -safe-string -g -I src/core/.qcheck_core.objs/byte -I src/core/.qcheck_core.objs/native -I /home/jmi/software/ocaml-11-09-2023/_opam/lib/bytes -I /home/jmi/software/ocaml-11-09-2023/_opam/lib/ocaml/unix -intf-suffix .ml -no-alias-deps -o src/core/.qcheck_core.objs/native/qCheck2.cmx -c -impl src/core/QCheck2.ml)
File "src/core/QCheck2.ml", line 1:
Error: Could not find the .cmi file for interface src/core/QCheck2.ml.
(cd _build/default && /home/jmi/software/ocaml-11-09-2023/_opam/bin/ocamlc.opt -w -40 -w +a-4-42-44-48-50-58-32-60@8 -safe-string -g -bin-annot -I src/core/.qcheck_core.objs/byte -I /home/jmi/software/ocaml-11-09-2023/_opam/lib/bytes -I /home/jmi/software/ocaml-11-09-2023/_opam/lib/ocaml/unix -intf-suffix .ml -no-alias-deps -o src/core/.qcheck_core.objs/byte/qCheck2.cmo -c -impl src/core/QCheck2.ml)
File "src/core/QCheck2.ml", line 1172, characters 2-20:
1172 |   fun_gen: 'b Gen.t; (** How to generate output values *)
         ^^^^^^^^^^^^^^^^^^
Warning 69 [unused-field]: record field fun_gen is never read.
(However, this field is used to build or mutate values.)

File "src/core/QCheck2.ml", line 1570, characters 4-33:
1570 |     mutable res: 'a TestResult.t;
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 69 [unused-field]: mutable record field res is never mutated.

```

**Comment by lthls** [comment]
It looks like the intended behaviour: the interface for `QCheck2.ml` generated by `dune` is named `qCheck2.cmi` instead of `QCheck2.cmi`, so with the new handling of filenames the compiler ignores it.
I think `dune` should be patched so that it stops generating output filenames that do not match the input.

**Comment by gasche** [comment]
cc @Octachron 

(If there is something easy we can do on the compiler side to preserve compatibility, I think it would be a good idea, at least during a transition period.)

**Comment by jmid** [comment]
FYI, the bytecode build (and hence the 32-bit build) continue to work, creating an unfortunate mismatch between the native and bytecode backends.

**Comment by lthls** [comment]
> FYI, the bytecode build (and hence the 32-bit build) continue to work, creating an unfortunate mismatch between the native and bytecode backends.

That is (again) because `dune` compiles the bytecode and native versions differently.
For implementations without an interface, the bytecode version is compiled normally, producing a `.cmo` and a `.cmi`. However, to prevent issues such as races to produce the `.cmi` or possible digest incompatibilities, `dune` compiles the native version with the flag `-intf-suffix .ml`. In normal cases, this makes the compiler check the implementation against the `.cmi` file produced by the bytecode compiler. That's where producing outputs with the wrong capitalization causes issues.

**Comment by Octachron** [comment]
In this case, this is an unintended restriction of the cmi lookup rules in #12389 : #12389 restricted the cmi lookup to `path/xy.cmi` (in the load paths). The fix in #12545 restores the previous behavior by lookup for `xy.cmi` (or `Xy.cmi`)  anywhere in the load paths.

**Comment by jmid** [comment]
Great, I can confirm that #12545 fixes it for me, thanks!  :+1:

**Comment by gasche** [comment]
Closed by merging #12545. I remain impressed by the versatility of property-based testing.
